pkg/csi/service/vanilla/controller.go

```diff
@@ -221,52 +226,73 @@ func (c *controller) Init(config *cnsconfig.Config, version string) error {
 	if err != nil {
-		log.Errorf("failed to read config. Error: %+v", err)
-		return
+		msg := fmt.Sprintf("failed to read config. Error: %+v", err)
+		log.Error(msg)
+		return errors.New(msg)
 	}
 	newVCConfig, err := cnsvsphere.GetVirtualCenterConfig(ctx, cfg)
 	if err != nil {
 		log.Errorf("failed to get VirtualCenterConfig. err=%v", err)
-		return
+		return err
 	}
 	if newVCConfig != nil {
 		var vcenter *cnsvsphere.VirtualCenter
 		if c.manager.VcenterConfig.Host != newVCConfig.Host ||
 			c.manager.VcenterConfig.Username != newVCConfig.Username ||
 			c.manager.VcenterConfig.Password != newVCConfig.Password {
-			log.Debugf("Unregistering virtual center: %q from virtualCenterManager", c.manager.VcenterConfig.Host)
-			err = c.manager.VcenterManager.UnregisterAllVirtualCenters(ctx)
-			if err != nil {
-				log.Errorf("failed to unregister vcenter with virtualCenterManager.")
-				return
+
+			// Verify if new configuration has valid credentials by connecting to vCenter.
+			// Proceed only if the connection succeeds, else return error.
+			newVC := &cnsvsphere.VirtualCenter{Config: newVCConfig}
+			if err = newVC.Connect(ctx); err != nil {
+				msg := fmt.Sprintf("failed to connect to VirtualCenter host: %s using new credentials, Err: %+v", newVCConfig.Host, err)
+				log.Error(msg)
+				return errors.New(msg)
 			}
-			log.Debugf("Registering virtual center: %q with virtualCenterManager", newVCConfig.Host)
-			vcenter, err = c.manager.VcenterManager.RegisterVirtualCenter(ctx, newVCConfig)
+
+			// Reset virtual center singleton instance by passing reload flag as true
+			log.Info("Obtaining new vCenterInstance using new credentials")
+			vcenter, err = cnsvsphere.GetVirtualCenterInstance(ctx, &cnsconfig.ConfigurationInfo{Cfg: cfg}, true)
 			if err != nil {
-				log.Errorf("failed to register VC with virtualCenterManager. err=%v", err)
-				return
+				msg := fmt.Sprintf("failed to get VirtualCenter. err=%v", err)
+				log.Error(msg)
+				return errors.New(msg)
 			}
-			c.manager.VcenterManager = cnsvsphere.GetVirtualCenterManager(ctx)
 		} else {
-			vcenter, err = c.manager.VcenterManager.GetVirtualCenter(ctx, newVCConfig.Host)
+			// If it's not a VC host or VC credentials update, same singleton instance can be used
+			// and it's Config field can be updated
+			vcenter, err = cnsvsphere.GetVirtualCenterInstance(ctx, &cnsconfig.ConfigurationInfo{Cfg: cfg}, false)
 			if err != nil {
-				log.Errorf("failed to get VirtualCenter. err=%v", err)
-				return
+				msg := fmt.Sprintf("failed to get VirtualCenter. err=%v", err)
+				log.Error(msg)
+				return errors.New(msg)
 			}
 			vcenter.Config = newVCConfig
 		}
 		c.manager.VolumeManager.ResetManager(ctx, vcenter)
-		c.manager.VolumeManager = cnsvolume.GetManager(ctx, vcenter)
 		c.manager.VcenterConfig = newVCConfig
+		c.manager.VolumeManager = cnsvolume.GetManager(ctx, vcenter)
+		// Re-Initialize Node Manager so topology feature gets latest vCenter config
+		c.nodeMgr = &Nodes{}
```

Comment:
`so topology feature gets latest vCenter config` You can remove this. Since this is not specific to topology feature. Even regular ControllerPublish and ControllerUnPublish workflows need the new node manager.

Reply:
updated comment as 

```
// Re-Initialize Node Manager to cache latest vCenter config
```